astropy/io/fits/hdu/hdulist.py

```diff
@@ -194,25 +273,30 @@ def __getitem__(self, key):
             hdus = super(HDUList, self).__getitem__(key)
             return HDUList(hdus)
 
-        idx = self.index_of(key)
-        return super(HDUList, self).__getitem__(idx)
+        # Originally this used recursion, but hypothetically an HDU with
+        # a very large number of HDUs could blow the stack, so use a loop
+        # instead
+        return self._try_while_unread_hdus(super(HDUList, self).__getitem__,
```

Comment:
Somehow this is where things go wrong on Windows. If you do:

```python
super(HDUList, self).__getitem__(self._positive_index_of(key))
```

then it fails when ``key`` is 2 (in the failing test). However, once the method is passed to ``_try_while_unread_hdus``, it apparently still runs and returns a ``PrimaryHDU`` for ``key=2``, which is very confusing.